### PR TITLE
Plugins: Expose "query" plugin metadata

### PR DESCRIPTION
**Pull request by cameronwaterman**
**What is this feature?**

Exposes a new plugin metadata field "query" that allows data source plugins to only be displayed in the dashboard query variable editor.

**Why do we need this feature?**

This is functionality that we've needed on several occasions and doesn't seem to be present in the current feature set.

**Who is this feature for?**

Developers of data source plugins.

**Which issue(s) does this PR fix?**:

N/A

**Special notes for your reviewer:**

Could we backport this change to a 9.X release?

Please check that:
- [X] It works as expected from a user's perspective.
- [X] If this is a pre-GA feature, it is behind a feature toggle.
- [X] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Comment by cameronwaterman** [comment]
"query" isn't a great field name, I'm open to suggestions here 🙂

**Comment by cameronwaterman** [comment]
@mikkancso @Eve832 @josmperez 

I'm seeing some unexpected behavior where the data source marked as "query" (i.e. only shows up in the query variable editor) is still showing up as the default data source in Explore rows (but isn't a selectable option in the DataSourcePicker). I wasn't able to find where this populates in source - is it possible there's some caching going on that is causing this?

image

**Comment by ryantxu** [comment]
Is there a data source that does not support query?

**Comment by cameronwaterman** [comment]
> Is there a data source that does not support query?

This adds a plugin metadata field that makes it so a data source is _only_ exposed as a query variable and not in other places.

**Comment by ryantxu** [comment]
I get what it does -- I just do not understand/know of a use case for a datasource that *only* can be used query variables.

The alternative is to add a query editor that says "this datasource can only be used to query variables" -- but also seems like showing the variables in a table is useful enough 🤷🏻

**Comment by sunker** [comment]
Thanks for creating this PR @cameronwaterman. I think this could make sense, however a change like this needs to start with a [feature request issue](https://github.com/grafana/grafana/issues/new?assignees=&labels=type%2Ffeature-request&projects=&template=1-feature_requests.md) so we can collect feedback from the community. Can you please go ahead and create one? 

In the meantime - one thing that might be helpful for your users is to skip defining a query editor in your plugin's `module.ts`. 
Screenshot from 2023-09-27 09-42-01

**Comment by cameronwaterman** [comment]
@sunker - Thanks for your feedback, we've already added the empty query editor to our data source. I've gone ahead and create the feature issue here: https://github.com/grafana/grafana/issues/75551

File: docs/sources/developers/plugins/metadata.md [hunk 1]
```diff
@@ -46,6 +46,7 @@ The plugin.json file is required for all plugins. When Grafana starts, it scans
...
 | `tracing`            | boolean                       | No       | For data source plugins, if the plugin supports tracing. Used for example to link logs (e.g. Loki logs) with tracing plugins.                                                                                                                                                                                                                                                                           |
+| `query`              | boolean                       | No       | For data source plugins, used to enable the plugin in the query variable editor.                                                                                                                                                                                                                                                                                                                        |
```
Comment (josmperez) [inline docs/sources/developers/plugins/metadata.md:49]:
```suggestion
| `query`              | boolean                       | No       | For data source plugins, enables the plugin in the query variable editor.                                                                                                                                                                                                                                                                                                                        |
```

**Review by josmperez — APPROVED** [review]
Added one nit suggestion. There's not much documentation impact, but what's there LGTM from a doc standpoint.

**Comment by sunker** [comment]
> @sunker - Thanks for your feedback, we've already added the empty query editor to our data source. I've gone ahead and create the feature issue here: #75551

Thanks @cameronwaterman. Let's wait for a while and give the community comment on your FR.

**Pull request closed without merging**